tests/subsys/mgmt/mcumgr/os_mgmt_echo/testcase.yaml

```diff
@@ -0,0 +1,9 @@
+#
+# Copyright (c) 2022 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+tests:
+  os.mgmt.echo:
+    platform_allow: qemu_cortex_m0 qemu_cortex_m3 nrf52840dk_nrf52840 nrf52dk_nrf52832 native_posix native_posix_64
```

Comment:
any reason why this test cannot run on other platforms?

Reply:
I started off with a `platform_disallow` list instead and found quite a few boards it would fail to compile on, e.g. leon, and platform_disallow is rejected by the compliance checks so had to switch to platform_allow

Reply:
> platform_disallow is rejected by the compliance checks

probably because `platform_disallow` does not exist; it is supposed to be `platform_exclude`.

> found quite a few boards it would fail to compile on, e.g. leon

but are they supposed to fail?

Reply:
I'll retry with platform_exclude... They fail because of not having a `sys_reboot()` implementation, which means os_mgmt cannot work

Reply:
Following boards fail to build because of lack of a reboot function:
```
qemu_nios2
qemu_malta_be
qemu_malta
qemu_leon3
qemu_cortex_a53_smp
qemu_cortex_a53
xenvm
xenvm_gicv3
qemu_xtensa
qemu_arc_hs6x
qemu_arc_hs
qemu_arc_em
nxp_ls1046ardb_smp_4cores
nxp_ls1046ardb_smp_2cores
nxp_ls1046ardb
emsdp_em9d
emsdp_em7d_esp
emsdp_em7d
emsdp_em6
emsdp_em5d
emsdp_em4
emsdp
em_starterkit_em7d_v22
em_starterkit_em7d
em_starterkit_em11d
em_starterkit
bcm958402m2_a72
altera_max10
mimx8mp_evk_a53
mimx8mp_evk_a53_smp
mimx8mn_evk_a53_smp
mimx8mn_evk_a53
mimx8mm_evk_a53_smp
mimx8mm_evk_a53
intel_socfpga_agilex_socdk
iotdk
intel_adsp_cavs25_tgph
intel_adsp_cavs25
intel_adsp_cavs20_jsl
intel_adsp_cavs20
intel_adsp_cavs18
intel_adsp_cavs15
hsdk_2cores
hsdk
gr716a_mini
generic_leon3
fvp_baser_aemv8r_smp
fvp_baser_aemv8r
fvp_base_revc_2xaemv8a
nsim_sem_mpu_stack_guard
nsim_sem
nsim_hs_smp
nsim_hs_mpuv6
nsim_hs
nsim_hs6x
nsim_hs6x_smp
nsim_em7d_v22
nsim_em
```
I can make a `platform_disallow` list of the above for this instead if wanted?

Reply:
So the problem is that `sys_arch_reboot` is not implemented by some architectures (most archs provide at least a no-op implementation that is defined as a weak symbol).

IMO, that should be considered a bug because the `sys_reboot` interface is not supposed to be optional (at least, according to the current definition -- i.e. `CONFIG_REBOOT` does not depend on anything).

If `sys_reboot` (or `sys_arch_reboot` for that matter) is supposed to be an optional feature, then it should provide `HAS_REBOOT` or something equivalent, which each arch/SoC implementation can select when they provide the `sys_arch_reboot` implementation, and the `os_mgmt` can depend on.

If `sys_reboot` is not supposed to be an optional feature, then `sys_arch_reboot` must be implemented by all architectures (and there ought to be a test that validates this).

p.s. In general, the reboot infrastructure seems to not have been very well thought out, and should be reworked at some point. As far as this PR goes, I would add `arch_exclude:` for the affected architectures (e.g. arm64, nios2, ...) and add a `FIXME:` above saying these archs are excluded because they are missing `sys_arch_reboot` implementation.

Reply:
Some SoCs (or rather, I have encountered a few soft-cores that) do not support software-rebooting. I think we need to make/keep the reboot functionality optional.

Reply:
> Does anyone know/remember why this was not done? cc @nordic-krch

As a no-op? Wouldn't that just lead to run-time failures when people use the interface (requested reboot ending up as a no-op)?

Reply:
> As a no-op? Wouldn't that just lead to run-time failures when people use the interface (requested reboot ending up as a no-op)?

Yes, I suppose the intention was to throw sort of an "unimplemented exception" (i.e. `Failed to reboot: spinning endlessly...` is displayed) when a SoC does not provide its own implementation:

https://github.com/zephyrproject-rtos/zephyr/blob/fb60aab2458684f347533cba3658c89757580932/arch/riscv/core/reboot.c#L16-L27

https://github.com/zephyrproject-rtos/zephyr/blob/fb60aab2458684f347533cba3658c89757580932/arch/arm/core/aarch32/cortex_a_r/reboot.c#L24-L27

But yes, this would be handled much better through something like `ARCH_HAS_REBOOT`.